### PR TITLE
Contrastive loss term

### PR DESCRIPTION
**Pull request by gabrieltseng**
See here for motivation: https://docs.google.com/presentation/d/1zeDlEqFMxBhOujxtFMDDfKEe6v07n2HpDhNWKT-fgv8/edit?usp=sharing

### CODE REVIEW

File: helios/train/train_module/galileo.py [hunk 1]
```diff
@@ -236,30 +244,41 @@ def train_batch(
...
+                        get_local_tensor((reg_term_a + reg_term_b) / 2)
+                        / num_microbatches
                     )
 
-                    # Run Encoder and decoder on the augmented input
-                    loss, latent, decoded, target_output = self.model_forward_b(
-                        masked_batch, patch_size, self.token_exit_cfg_b
+                if self.contrastive_loss is not None:
+                    contrastive_loss = self.contrastive_loss.compute(latent_a, latent_b)
+                    loss += contrastive_loss
```
Comment (gabrieltseng) [inline helios/train/train_module/galileo.py:281]: Strictly speaking there should be a linear layer between the pooled latents and the contrastive loss. See the CROMA paper below

Screenshot 2025-04-08 at 12 50 27 PM

File: helios/train/train_module/galileo.py [hunk 2]
```diff
@@ -236,30 +244,41 @@ def train_batch(
...
-                    )
+                loss_a, latent_a = self.apply_masks_and_compute_losses_and_latents(
+                    microbatch,
+                    self.masking_strategy_a.apply_mask,
+                    self.model_forward_a,
+                    patch_size,
+                    self.token_exit_cfg_a,
+                )
+                loss_b, latent_b = self.apply_masks_and_compute_losses_and_latents(
+                    microbatch,
```
Comment (gabrieltseng) [inline helios/train/train_module/galileo.py]: This applies different masks to the same microbatch, which better mirrors Galileo.

Also it might speed training up? Since we get more learning signal for the same data loading cost.

File: helios/train/train_module/galileo.py (hunk 1 above)
Comment (gabrieltseng) [reply helios/train/train_module/galileo.py:281]: added: https://wandb.ai/eai-ai2/helios-debug/runs/o38klaxi?nw=nwusergabrieltseng

Comment (pjreddie) [reply helios/train/train_module/galileo.py:281]: Now you run the model twice per batch with both masking strategies etc instead of alternating right? Are we at all worried about backward compatibility since this changes from previous experiments? Probably doesn't matter really though. Is this how Galileo (NASA version) works?

File: helios/train/loss.py [hunk 3]
```diff
@@ -514,6 +514,52 @@ def compute(
...
+class InfoNCELoss(Loss):
+    """Loss function for InfoNCE."""
+
+    name = "InfoNCE"
+
+    def __init__(
+        self,
+        tau: float = 0.1,
+    ):
+        """Initialize adjusted patch discrimination loss.
```
Comment (pjreddie) [inline helios/train/loss.py]: Nit: probably should be `"""Initialize InfoNCE loss"""`

File: helios/train/loss.py
Comment (pjreddie) [inline helios/train/loss.py:1]: probably should add a test for the new loss

File: helios/train/train_module/galileo.py (hunk 1 above)
Comment (gabrieltseng) [reply helios/train/train_module/galileo.py:281]: yes. it does affect compatibility but not massively i think (?)

and sort of - Galileo did use this approach (multiple re-uses of a same minibatch) but (i) did it 4 times, and (ii) subsampled afresh each time.

File: helios/train/loss.py
Comment (gabrieltseng) [reply helios/train/loss.py:1]: added!
